pkg/error_code/error_code.go

```diff
@@ -31,42 +36,117 @@
 package errcode
 
 import (
+	"fmt"
 	"net/http"
+	"strings"
 )
 
-// RegisteredCode helps document that we are using a registered error code that must never change
-type RegisteredCode string
-
-const (
-	// InternalCode means the operation placed the system is in an inconsistent or unrecoverable state.
-	// Essentially a handled panic.
-	// This is exactly the same as a HTTP 500, so it is not necessary to send this code over HTTP.
-	InternalCode RegisteredCode = "internal"
-	// InvalidInputCode is a validation failure of an input.
-	// The response will indicate the exact input issue.
-	InvalidInputCode RegisteredCode = "input"
-	// NotFoundCode indicates a resource is missing.
-	// The response can indicate what resource is missing.
-	// However, when a single resource is requested, that may be implicit.
-	NotFoundCode RegisteredCode = "missing"
+// CodeStr is a representation of the type of a particular error.
+// The underlying type is string rather than int.
+// This enhances both extensibility (avoids merge conflicts) and user-friendliness.
+// A CodeStr can have dot separators indicating a hierarchy.
+type CodeStr string
+
+func (str CodeStr) String() string { return string(str) }
+
+// A Code has a CodeStr representation.
+// It is attached to a Parent to find metadata from it.
+// The Meta field is provided for extensibility: e.g. attaching HTTP codes.
+type Code struct {
+	// CodeStr contains the full dot-separted path.
+	// This allows for equality comparison based on CodeStr.
+	CodeStr CodeStr
+	Parent  *Code
+}
+
+// NewCode create a new top-level code.
+// A top-level code must not contain any dot separators: that will panic
+// Most codes should be created from hierachy with the Child method.
+func NewCode(codeRep CodeStr) Code {
+	code := Code{CodeStr: codeRep}
+	if err := code.checkCodePath(); err != nil {
+		panic(err)
+	}
+	return code
+}
+
+// Child creates a new code from a parent.
+// childStr must include the parent codes with dot-separation for documentation purposes.
+// An incorrect parent reference in the string panics.
+func (code Code) Child(childStr CodeStr) Code {
+	child := Code{CodeStr: childStr, Parent: &code}
+	if err := child.checkCodePath(); err != nil {
+		panic(err)
```

Comment:
I agree. However, panic at initialization time is inevitable in go and not really a problem. One does not just compile without testing and ship a program in go. panics that occur based on conditional logic or otherwise don't automatically happen at startup are the issue.